src/atmega2560p/hal/port.rs

```diff
@@ -0,0 +1,185 @@
+    port: *mut Port,
+    pin: usize,
+}
+
+/// Type ```IOMode```
+/// Represents the Input/Output mode of the pin.
+pub enum IOMode {
+    Input,
+    Output,
+}
+
+impl Port {
+    /// Returns mutable reference to the Port of given PortName.
+    pub fn new(name: PortName) -> &'static mut Port {
+        match name {
+            PortName::A => unsafe { &mut *(0x20 as *mut Port) },
+            PortName::B => unsafe { &mut *(0x23 as *mut Port) },
+            PortName::C => unsafe { &mut *(0x26 as *mut Port) },
+            PortName::D => unsafe { &mut *(0x29 as *mut Port) },
+            PortName::E => unsafe { &mut *(0x2C as *mut Port) },
+            PortName::F => unsafe { &mut *(0x2F as *mut Port) },
+            PortName::G => unsafe { &mut *(0x32 as *mut Port) },
+            PortName::H => unsafe { &mut *(0x100 as *mut Port) },
+            PortName::J => unsafe { &mut *(0x103 as *mut Port) },
+            PortName::K => unsafe { &mut *(0x106 as *mut Port) },
+            PortName::L => unsafe { &mut *(0x109 as *mut Port) },
+        }
+    }
+
+    /// Returns PortName of port of the given address input.
+    /// Panics if the address is invalid.
+    pub fn name(&self) -> PortName {
+        let address = (self as *const Port) as usize; // Gets address of port.
+        match address {
+            //  Return PortName based on the address read.
+            0x20 => PortName::A,
+            0x23 => PortName::B,
+            0x26 => PortName::C,
+            0x29 => PortName::D,
+            0x2C => PortName::E,
+            0x2F => PortName::F,
+            0x32 => PortName::G,
+            0x100 => PortName::H,
+            0x103 => PortName::J,
+            0x106 => PortName::K,
+            0x109 => PortName::L,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Returns a ```Some<Pin>``` if pin number is valid and returns none if not valid.
+    pub fn pin(&mut self, pin: usize) -> Option<Pin> {
+        if pin < 0x8 {
+            Some(Pin { port: self, pin })
+        } else {
+            None
+        }
+    }
+}
+
+impl Pin {
+    ///Return a pin for the given port name and pin number.
+    pub fn new(port: PortName, pin: usize) -> Option<Pin> {
+        Port::new(port).pin(pin)
+    }
+
+    /// Change pin mode to input or output by changing the DDr register.
+    pub fn set_pin_mode(&mut self, mode: IOMode) {
+        //  Read the value of DDxn register.
+        let mut ddr_val = unsafe { read_volatile(&mut (*self.port).ddr) };
+
+        //  Calculate the value to be written to DDxn register.
+        ddr_val &= !(0x1 << self.pin);
+        ddr_val |= match mode {
+            IOMode::Input => 0x0,
+            IOMode::Output => 0x1 << self.pin,
+        };
+
+        // Write the value to DDxn register.
+        unsafe { write_volatile(&mut (*self.port).ddr, ddr_val) }
+    }
+
+    /// Toggles the value of PORTxn by writing one to PINxn ,independent of the value of DDRxn.
+    pub fn toggle(&mut self) {
```

Comment:
Again we don't need what happens inside, I need to know what toggle does for me.

Reply:
fixed